src/qiskit_qec/circuits/code_circuit.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
+
+"""Class that manage circuits for codes."""
+from abc import ABC, abstractmethod
+
+
+class CodeCircuit(ABC):
+    """
+    Abstract class to manage circuits for codes, as well
+    as other fault-tolerant circuits.
+
+    A CodeCircuit requires the methods `string2nodes`,
+    `check_nodes` and `is_cluster_neutral` in order to
+    interface with its `DecodingGraph` and decoders.
+    """
+
+    def __init__(self):
+        """
+        Initialization of classes that inherent from CodeCircuit can
+        be done in various ways, depending on the code or code family
+        to be initialized. In all cases, the initialization must define
+        attributes `circuit` and `base`. The former is a dictionary with
+        circuits as values, and labels as keys. The latter is the label
+        regarded as the base case, used in decoding graph generation.
+        """
+        pass
+
+    @abstractmethod
+    def string2nodes(self, string, **kwargs):
+        """
+        Convert output string from circuits into a set of nodes for
+        `DecodingGraph`.
+        Args:
+            string (string): Results string to convert.
+            kwargs (dict): Any additional keyword arguments.
+        """
+        pass
+
+    @abstractmethod
+    def check_nodes(self, nodes, ignore_extra_boundary=False):
```

Comment:
Same thing as above: dynamically computed by default, but can be overwritten by classes that inherit from it for more custom behavior.

Reply:
I'm not sure how this could be done entirely generally. This could be a project to save for your PhD!